feat(searchbar): Include a short explaination in searchbar

wmak opened the pull request:
- This adds a short explanation for functions & fields in
  discover&dashboards
  - Explanations are intentionally kept short since we don't have much space here
- This sorts the tag keys so they appear in a consistent order


### Preview
<img width="867" alt="image" src="https://user-images.githubusercontent.com/4205004/164786606-1f0a4ab2-9938-4ce5-96f4-c37d6ca90c6b.png">
<img width="867" alt="image" src="https://user-images.githubusercontent.com/4205004/164786630-b3cc8199-76cf-42c3-be8c-d77ec7054015.png">

wmak committed:
searchbar: Include a short explaination in searchbar

- This adds a short explaination for functions & fields in
  discover&dashboards
- This sorts the tag keys so they appear in a consistent order

narsaynorath commented on static/app/utils/discover/fields.tsx:
```diff
@@ -1045,6 +1073,106 @@ export function isNumericMetrics(field: string): boolean {
+  [FieldKey.ID]: 'event identifier',
+  [FieldKey.ISSUE]: 'issue short id',
+  [FieldKey.LEVEL]: 'string',
+  [FieldKey.LOCATION]: 'where error happend',
```
```suggestion
  [FieldKey.LOCATION]: 'where error happened',
```

narsaynorath commented on static/app/utils/discover/fields.tsx:
```diff
@@ -1045,6 +1073,106 @@ export function isNumericMetrics(field: string): boolean {
+  [FieldKey.LOCATION]: 'where error happend',
+  [FieldKey.MESSAGE]: 'title or name',
+  [FieldKey.OS_BUILD]: 'internal build revision',
+  [FieldKey.OS_KERNEL_VERSION]: 'kernal string',
```
```suggestion
  [FieldKey.OS_KERNEL_VERSION]: 'kernel string',
```

narsaynorath commented on static/app/components/smartSearchBar/index.tsx:
```diff
@@ -751,7 +755,13 @@ class SmartSearchBar extends React.Component<Props, State> {
+        .map(value => ({
+          value,
+          desc: value,
+          documentation: getFieldDoc ? getFieldDoc(value.slice(0, -1)) : '',
```
```suggestion
          documentation: getFieldDoc?.(value) ?? '',
```
I think this should be valid too, if `getFieldDoc` is undefined it won't call it, and I don't think we need to copy the string since it's a primitive

wmak replied:
whoa fancy, didn't realize i could do this :+1:

wmak replied:
oh actually this isn't copying the string, its stripping the last character cause the keys actually come as
`count():` (ie. with the extra `:` at the end)

narsaynorath commented on static/app/components/smartSearchBar/index.tsx:
```diff
@@ -751,7 +755,13 @@ class SmartSearchBar extends React.Component<Props, State> {
+          desc: value,
+          documentation: getFieldDoc ? getFieldDoc(value.slice(0, -1)) : '',
+        }))
+        .sort((a, b) => a.value > b.value),
```
Look like we use `localeCompare` for these kinds of sorts, does this work?
```suggestion
        .sort((a, b) => a.value.localeCompare(b.value),
```

narsaynorath reviewed (commented):
Just some comments to start with, looking good but I'm going to poke around locally

shruthilayaj commented on static/app/utils/discover/fields.tsx:
```diff
@@ -1045,6 +1073,106 @@ export function isNumericMetrics(field: string): boolean {
 }
 
+export const FIELDS_DOCS: Readonly<Record<string, string>> = {
+  has: 'check field exists',
```
do we need to localize all these docstrings?
```suggestion
  has: t('check field exists'),
```

evanpurkhiser commented on static/app/components/smartSearchBar/index.tsx:
```diff
@@ -151,6 +151,10 @@ type Props = WithRouterProps & {
+  /**
+   * A function to get documentation for a field
+   */
+  getFieldDoc?: (key: string) => string;
```
How about `React.ReactNode` so we can bold things etc etc maybe?

evanpurkhiser commented on static/app/utils/discover/fields.tsx:
```diff
@@ -134,6 +134,7 @@ const CONDITIONS_ARGUMENTS: SelectValue<string>[] = [
 export const AGGREGATIONS = {
   count: {
     parameters: [],
+    documentation: 'number of events',
```
These all need `t()`

wmak committed:
address pr comments

- wrap the docs in t()
- spelling mistake
- clean up the sort a bit
- return ReactNode instead of just string

wmak committed:
typing fix

narsaynorath commented on static/app/components/smartSearchBar/searchDropdown.tsx:
```diff
@@ -195,3 +196,10 @@ const Description = styled('span')`
+const Documentation = styled('span')`
+  font-size: ${p => p.theme.fontSizeSmall};
+  font-family: ${p => p.theme.text.familyMono};
+  float: right;
```
![Screen Shot 2022-04-22 at 5 00 29 PM](https://user-images.githubusercontent.com/22846452/164792810-181f73eb-38d3-47f2-abfa-c13afdf825bd.png)

lol oops that causes this on the issues page. I think we can still use the above suggestion but wrap the `item.title` and `Description` bit in some wrapper and it'll work :eyes: maybe give it a shot, or put it in as a separate PR?

evanpurkhiser replied:
Please no `float` 😢

wmak replied:
Hmm what's better to use here instead of `float`?

priscilawebdev replied:
we should avoid suing ellipses

priscilawebdev replied:
very good feedback @narsaynorath 👏

wmak replied:
Going to let y'all take a look at the design of this in a different PR so I can get this feature up for now

wmak committed:
ref: move the couldnt do anything return to the end

wmak committed:
change expected since we're sorting now

priscilawebdev commented on static/app/utils/discover/fields.tsx:
```diff
@@ -240,6 +248,7 @@ export const AGGREGATIONS = {
         required: true,
       },
     ],
+    documentation: t('minimum'),
```
hm this is more like a description 🧐

priscilawebdev commented:
the experience on smaller devices looks a bit odd:

![image](https://user-images.githubusercontent.com/29228205/165044211-2b7e1968-431b-4820-bf1a-fce4fb47da99.png)


should we have two columns and break the text?
